File: SortVision/public/code/bucket/csharp/bucketSort.cs
using System;
using System.Collections.Generic;

public static class BucketSort {
    // Bucket Sort (for floats in [0, 1))
    // Time: Avg O(n + k), Worst O(n^2), Space: O(n + k)
    public static void Sort(float[] arr) {
        if (arr == null || arr.Length <= 1) return;

        int n = arr.Length;
        var buckets = new List<float>[n];
        for (int i = 0; i < n; i++) buckets[i] = new List<float>();

        foreach (float x in arr) {
            int idx = Math.Min(n - 1, (int)(x * n));
            buckets[idx].Add(x);
        }

        int k = 0;
        foreach (var bucket in buckets) {
            bucket.Sort();
            foreach (float x in bucket) arr[k++] = x;
        }
    }

    public static void Main() {
        float[] arr = {0.42f, 0.32f, 0.23f, 0.52f, 0.25f, 0.47f, 0.51f};
        Sort(arr);
        Console.WriteLine(string.Join(", ", arr));
    }
}

File: SortVision/public/code/bucket/dart/bucketSort.dart
class BucketSort {
  // Bucket Sort (for floats in [0, 1))
  // Time: Avg O(n + k), Worst O(n^2), Space: O(n + k)
  static void sort(List<double> arr) {
    if (arr.length <= 1) return;

    final n = arr.length;
    final buckets = List.generate(n, (_) => <double>[]);

    for (final x in arr) {
      final idx = (x * n).floor().clamp(0, n - 1);
      buckets[idx].add(x);
    }

    int k = 0;
    for (final bucket in buckets) {
      bucket.sort();
      for (final x in bucket) arr[k++] = x;
    }
  }
}

void main() {
  final arr = [0.42, 0.32, 0.23, 0.52, 0.25, 0.47, 0.51];
  BucketSort.sort(arr);
  print(arr);
}
